### DISCUSSION TITLE
How do I create a remote workspace directly?

### DISCUSSION THREAD

**Discussion by eliaskickbush** [pulumi/pulumi#9032] [Automation API] [answered]
Hello community, as part of using the automation API, I find myself needing to list all of my currently existing stacks.

For this I am using the Workspace method called ListStacks. However, in order to get a workspace that points to my remote backend (S3), I'm currently creating a dummy stack in order to use that stack's method to fetch the workspace:

```
ctx := context.Background()
stack, err := auto.UpsertStackInlineSource(ctx, "list-helper", "project", func(ctx *gopulumi.Context) error {
    return nil
})
if err != nil {
    fmt.Println("there was an error creating temporary stack: ", err.Error())
    return errors.New("could not create temporary stack")
}

wksp := stack.Workspace()
stacks, err := wksp.ListStacks(ctx)
if err != nil {
    fmt.Println("there was an error fetching workspace: ", err.Error())
    return errors.New("could not create temporary stack")
}
```

Isn't there a direct way to create a Workspace that points to the remote backend? Currently, the only way I can find to directly create a Workspace is to create a Local one, with `NewLocalWorkspace`.

Thanks in advance!

**Comment by EvanBoyle** [comment, marked as answer]
> Isn't there a direct way to create a Workspace that points to the remote backend?

The `ProjectSettings` determine what backend to connect to. Here is an example where the project settings specify a local file. This is not exactly what you are looking for, but this code could be updated so that the `Backend` points to an S3 filestate backend:

https://github.com/pulumi/automation-api-examples/blob/80afd678f6adc51eb9f0279ca62408b7f1e4623c/go/inline_passphrase_secrets_provider/main.go#L94-L100

There is some confusing and conflicting terminology in Automation API, so I'd like to define a few things to help clarify. 

In the context of Automation API and source, the terms `Remote` (source code in git), `Local` (source code in a folder on your machine), and `Inline` (function pointer in process) all refer to the location of the pulumi program source code, not the machine or backend where the code executes.

[`Workspace`](https://github.com/pulumi/pulumi/blob/013036064f8a49fca6842179a40e005da06a806a/sdk/go/auto/workspace.go#L27) is the execution context containing a Pulumi project. Today the only implementation of that interface is [`LocalWorkspace`](https://github.com/pulumi/pulumi/blob/33ed84b84ed08557e8bbc4519e81761caa57cce1/sdk/go/auto/local_workspace.go#L44) where the backing Pulumi CLI is on the machine running your program. The thought with this interface is that we might one day implement a different kind of workspace where the Pulumi CLI and machinery were running in the cloud, on a remote server, etc.
